driver-core/src/main/com/mongodb/internal/operation/ClientBulkWriteOperation.java

```diff
@@ -0,0 +1,1052 @@
+        MongoWriteConcernException writeConcernException = Exceptions.createWriteConcernException(
+                bulkWriteCommandOkResponse, connection.getDescription().getServerAddress());
+        if (writeConcernException != null) {
+            throw new MongoWriteConcernWithResponseException(writeConcernException, exhaustiveBulkWriteCommandOkResponse);
+        }
+        return exhaustiveBulkWriteCommandOkResponse;
+    }
+
+    private <R> R doWithRetriesDisabledForCommand(
+            final RetryState retryState,
+            final String commandDescription,
+            final Supplier<R> actionWithCommand) {
+        Optional<Boolean> originalRetryableCommandFlag = retryState.attachment(AttachmentKeys.retryableCommandFlag());
+        Supplier<String> originalCommandDescriptionSupplier = retryState.attachment(AttachmentKeys.commandDescriptionSupplier())
+                .orElseThrow(Assertions::fail);
+        try {
+            retryState.attach(AttachmentKeys.retryableCommandFlag(), false, true)
+                    .attach(AttachmentKeys.commandDescriptionSupplier(), () -> commandDescription, false);
+            return actionWithCommand.get();
+        } finally {
+            originalRetryableCommandFlag.ifPresent(value -> retryState.attach(AttachmentKeys.retryableCommandFlag(), value, true));
+            retryState.attach(AttachmentKeys.commandDescriptionSupplier(), originalCommandDescriptionSupplier, false);
+        }
+    }
+
+    private List<List<BsonDocument>> exhaustBulkWriteCommandOkResponseCursor(
+            final ConnectionSource connectionSource,
+            final Connection connection,
+            final BsonDocument response) {
+        int serverDefaultCursorBatchSize = 0;
+        try (BatchCursor<BsonDocument> cursor = cursorDocumentToBatchCursor(
+                TimeoutMode.CURSOR_LIFETIME,
+                response,
+                serverDefaultCursorBatchSize,
+                codecRegistry.get(BsonDocument.class),
+                options.getComment().orElse(null),
+                connectionSource,
+                connection)) {
+            return stream(spliteratorUnknownSize(cursor, ORDERED | IMMUTABLE), false).collect(toList());
+        }
+    }
+
+    private BsonDocumentWrapper<?> createBulkWriteCommand(
+            final boolean effectiveRetryWrites,
+            final WriteConcern effectiveWriteConcern,
+            final SessionContext sessionContext,
+            final List<? extends ClientNamespacedWriteModel> unexecutedModels,
+            final BatchEncoder batchEncoder,
+            final Runnable ifCommandIsRetryable) {
+        // BULK-TODO This implementation must limit the number of `models` it includes in a batch if needed.
+        // Each batch re-selects a server and re-checks out a connection because this is simpler and it is allowed,
+        // see https://mongodb.slack.com/archives/C035ZJL6CQN/p1722265720037099?thread_ts=1722264610.664109&cid=C035ZJL6CQN.
```

Comment:
Nit: These archives will be lost over time, so recommend adding a tldr here instead.

Reply:
Done in 14bb86ef3b832c743962f3cb1cb2c565284e3ec6. I moved the comment, and linked the relevant `DRIVERS` ticket there (the ticket will not be done, but it has the relevant discussion).